test/framework/src/main/java/org/elasticsearch/test/rest/section/ResponseBodyAssertion.java

```diff
@@ -0,0 +1,149 @@
+package org.elasticsearch.test.rest.section;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.test.rest.parser.RestTestFragmentParser;
+import org.elasticsearch.test.rest.parser.RestTestParseException;
+import org.elasticsearch.test.rest.parser.RestTestSuiteParseContext;
+
+/**
+ * Checks that the response body matches some text. 
+ */
+public class ResponseBodyAssertion extends Assertion {
+    public static final RestTestFragmentParser<ResponseBodyAssertion> PARSER = new RestTestFragmentParser<ResponseBodyAssertion>() {
+        @Override
+        public ResponseBodyAssertion parse(RestTestSuiteParseContext parseContext) throws IOException, RestTestParseException {
+            try (XContentParser parser = JsonXContent.jsonXContent.createParser(parseContext.parseField())) {
+                return new ResponseBodyAssertion("$body", parser.map());
+            }
+        }
+    };
+
+    private ResponseBodyAssertion(String field, Map<String, Object> expectedValue) {
+        super(field, expectedValue);
+    }
+
+    @Override
+    protected void doAssert(Object actualValue, Object expectedValue) {
+        if (false == expectedValue.equals(actualValue)) {
+            @SuppressWarnings("unchecked")
+            Map<String, Object> actual = (Map<String, Object>) actualValue;
+            @SuppressWarnings("unchecked")
+            Map<String, Object> expected = (Map<String, Object>) expectedValue;
+            FailureMessage message = new FailureMessage();
+            message.compareMaps(actual, expected);
+            throw new AssertionError(message.message);
+        }
+    }
+
+    private class FailureMessage {
```

Comment:
That is an interesting idea! I'd still have to do the recursive map/list walking stuff but instead of outputting error messages I'd be build matchers. As much as I like the error messages that come out of hamcrest, I wonder if they'd be as easy to read though.